### PR TITLE
SPARK-12729. PhantomReference to replace finalize in python broadcast…

### PR DESCRIPTION
**Pull request by GayathriMurali**
## What changes were proposed in this pull request?

Replace finalize() method in PythonBroadcast with Phantom Reference. Redesigned some portions of the code to better handle the thread to avoid thread leak. Fixed existing null pointer exceptions. Introduced a new thread class and a phantom reference class. 
## How was the this patch tested?

build/sbt "test-only org.apache.spark.api.python.*" - Passes all tests

### CODE REVIEW

**Comment by GayathriMurali** [comment]
@davies @zsxwing Can you please verify this patch?

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala [hunk 1]
```diff
@@ -871,6 +874,44 @@ private class PythonAccumulatorParam(@transient private val serverHost: String,
...
  */
 // scalastyle:off no.finalize
+
+/**
+ * Create a class that extends PhantomReference.
+ */
+private[spark] class FilePhantomReference(@transient var f: File, var q: ReferenceQueue[File])
+  extends PhantomReference(f, q){
+
+  private def cleanup()
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: who calls this method?

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala [hunk 2]
```diff
@@ -892,8 +933,18 @@ private[spark] class PythonBroadcast(@transient var path: String) extends Serial
   private def readObject(in: ObjectInputStream): Unit = Utils.tryOrIOException {
     val dir = new File(Utils.getLocalDir(SparkEnv.get.conf))
     val file = File.createTempFile("broadcast", "", dir)
+    val queue = new ReferenceQueue[File]()
+    val exitWhenFinished=false
+    val phantomReferences = new ListBuffer[FilePhantomReference]()
+    val threadName = "WeakReference"
+    val phantomThread=new PhantomThread(threadName,queue,phantomReferences)
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: it's not a good idea that creating a new thread for each PythonBroadcast.

**Comment by GayathriMurali** [comment]
@zsxwing  Most documentations on phantom reference i referred to suggests a separate daemon thread to do the cleanup. I can try adding the phantomreference removal from the queue in the same function without a thread. Would that work?

**Comment by zsxwing** [comment]
> @zsxwing Most documentations on phantom reference suggests a separate daemon thread to do the cleanup. I can try adding the phantom reference removal from the queue in the same function without a thread. Would that work?

The problem is it will create one thread for each PythonBroadcast. If a task has hundreds of `PythonBroadcast`s, it will create hundreds of threads.

**Comment by GayathriMurali** [comment]
@zsxwing how about creating one thread and the reference queue would contain all instances of the file object that needs to be GC'ed? Start a global thread?

**Comment by GayathriMurali** [comment]
@zsxwing  Should i modify the code to use one thread?

**Comment by zsxwing** [comment]
@GayathriMurali actually, I'm not sure if it will work. E.g., you need to create a global thread, but when to stop that?

**Comment by GayathriMurali** [comment]
@zsxwing when the queue is empty? Add a listener to the queue to invoke or stop the thread accordingly?  Would this approach still make Phantom reference more beneficial than finalize()?

**Comment by zsxwing** [comment]
> @zsxwing when the queue is empty? Add a listener to the queue to invoke or stop the thread accordingly? Would this approach still make Phantom reference more beneficial than finalize()?

Right. I'm thinking about it. Actually, it looks much more complicated than it was thought. Maybe let's just leave it until someone complaints the bad performance.

**Comment by srowen** [comment]
Why not use a finalizer? this is looking pretty complex with new threads and reference schemes otherwise

**Comment by GayathriMurali** [comment]
@srowen  The JIRA was created by @davies and the intent was to replace finalize() with Phantom Reference. http://resources.ej-technologies.com/jprofiler/help/doc/index.html

**Comment by srowen** [comment]
I think this is the wrong solution. `finalize()` should not do significant work unless something went wrong -- some expensive resource was not closed. Is that happening regularly? then I think that call site needs to be fixed. This pushes the problem somewhere else.

**Comment by davies** [comment]
The finalize() here is to cleanup the disk file for Python broadcasts, is the regular path. 

@srowen This change is requested by @rxin,  to avoid blocking operations in finalize().

Since the Python broadcasts are rarely used and exists() and delete() should be lightly system calls, I'm fine with current finalizer.

**Comment by srowen** [comment]
Another way to avoid blocking in `finalize` is to not do the work in the finalizer. It's as simple as making an idempotent "close" or "cleanup" method that does nothing if it has already been disposed. ... are these broadcasts not otherwise cleaned up?

To the extent they're rarely used, is this a problem?

My concern is that you're simply reimplementing a separate 'finalizer' queue with all the associated complexity.

**Comment by rxin** [comment]
The problem is blocking GC threads ...

**Comment by srowen** [comment]
Yes, I certainly understand that. `finalize` shouldn't do non-trivial work. Objects shouldn't require non-trivial cleanup, in general, when they hit the finalizer. This is typically a symptom of not cleaning up the object in normal execution paths. Why is that not possible?

**Comment by rxin** [comment]
Yea we could do that too. Actually explicit management is probably better, with a fallback to do implicit management so it is robust against mem leaks. That is provided somebody has the cycles to do it.

**Comment by rxin** [comment]
Thanks for the pull request. I'm going through a list of pull requests to cut them down since the sheer number is breaking some of the tooling we have. Due to lack of activity on this pull request, I'm going to push a commit to close it. Feel free to reopen it or create a new one. We can also continue the discussion on the JIRA ticket.

**Pull request closed without merging**